### ISSUE TITLE
[WIP] Optimize PHP files parsing

### ISSUE THREAD

**Issue by theofidry** [infection/infection#86]
I may have misread this but from what I see:

- we get the source files
- we generate the mutants for those files
- for each mutants, we parse the file, apply the mutation and run it

If that's what is happening, there I think we could optimize it as in the case where multiple mutants are applied to the same file, we will be parsing that file multiple times unless PhpParser has some kind of caching which I don't think it has.

**Comment by sidz** [comment]
@theofidry yup you're right. We parse one file so many times as mutations for one file we have + one time here: https://github.com/infection/infection/blob/master/src/Mutant/Generator/MutationsGenerator.php#L138

**Comment by theofidry** [comment]
Ok so it's not a mistake of mine, well we might have a nice possible performance improvement there then 😄

**Comment by sidz** [comment]
@theofidry if you'll share your thoughts about this improvement I'll be able to pick it up. If you don't mind :P

**Comment by theofidry** [comment]
Sorry I thought that was clear from the issue description. Basically the current behaviour is not ideal as soon as you have multiple mutants for the same file as you'll parse the file multiple times.

As this (having multiple mutants per file) is quite common, a possible optimisation would be to cache the result (the AST graph) of the parsed file. So parse the file only once.

IIRC there is no native way to clone an AST graph from PHP-Parser which is required in our case since we are going to modify it. But serialising it and unserialising it should work.

**Comment by sidz** [comment]
Yeah the idea was clear. Just wanted to know your thoughts about the best approach in this case.

Ah very nice. I thought about serialising today and present this idea to @borNfreee.

**Comment by theofidry** [comment]
As a first approach I would try to go for the simplest way, a simple array property with serialize/unserialize.

The other points, other way to cache it really add more complexity so we should make sure there is a good perf. improvement before going in that direction.

I also never saw anything about the serialize/unserialize for PHP-Parser but I heard that was the way to do it. It would be nice to have a confirmation though and if that's the case, either mention it in PHP-Parser docs and/or add a clone function in PHP-Parser to make it simpler

**Comment by borNfreee** [comment]
Another idea is to clone the mutated node in MutantCreator (see https://github.com/nikic/PHP-Parser/blob/e2e99f269b5b543c9196e1cf56348942165e8faf/lib/PhpParser/NodeVisitor/CloningVisitor.php) and then, in the next iteration, reuse `originalStatements` and restore original node from the updated AST before mutating.

Btw, I would suggest to create an issue in PHP-Parser lib to hear Nikic's opinion on reusing AST.
Just read all titles of all 250+ issue and didn't find a thread about it, but I have a feeling I saw it :|

**Comment by theofidry** [comment]
>Another idea is to clone the mutated Node [...]

@borNfreee I thought about this but I think that would create more issues than anything. I think trashing the AST is easier as otherwise you require to manage a restore step besides a mutate. But maybe the later will bring better performances

**Comment by sidz** [comment]
fyi: Nikic's thoughts https://twitter.com/plbSiD/status/935260598585217029

**Comment by sidz** [comment]
So with `serialize/unserialize` we save around 0.5-1 sec.

Here is code: https://github.com/sidz/infection/commit/e4217d2fa19c4719f1fdc891efb8c8d9c8fbe906

if we're happy with this performance boost I can prepare PR

**Comment by theofidry** [comment]
@sidz that looks interesting, I would love to see a Blackfire comparision to be able to assess more the perf. improvement but I realistically can't right now.

Note for the future, might be better to open an issue for it actually, it would be interesting to have a benchmarking script in the lib like I did for [alice](https://github.com/nelmio/alice/blob/master/Makefile#L43) to be able to reliable work on performances.

It doesn't need to be part of all the tests, but we could for example have a script clone a project of a specific version and run infection on it and profile that.

**Comment by borNfreee** [comment]
I did testing on your branch @sidz for JMS serializer project (~1400 mutatons), here are the results.

Since blacksfire's report is useless when xdebug is automatically disabled, I did comparisons with and without this new feature.

#### Without "disabling xdebug" feature:

№ of run | branch | time | memory usage | blackfire profile
--- | --- | --- | --- | ---
1 | optimize-php-file-parsing | 10m 56s | 214MB | https://blackfire.io/profiles/012f0704-0bf2-4438-8662-b796736010d5/graph
1 | master | 10m 51s | 200MB | https://blackfire.io/profiles/a71ce29b-4f82-4581-acea-59a32bc5053e/graph
2 | optimize-php-file-parsing |  10m 58s | 214MB | https://blackfire.io/profiles/ac575fea-2f9d-490c-b690-f7b8ae27e9f4/graph
2 | master | 10m 47s | 200MB | https://blackfire.io/profiles/d216a715-22c9-409b-adf7-a194acb196c6/graph

Comparing profile (from `optimize-php-file-parsing` to `mater`): https://blackfire.io/profiles/compare/c812ebc6-47b7-40fa-8e68-0e0dcc6430da/graph

#### With "disabling xdebug" feature:

№ of run | branch | time | memory usage | blackfire profile
--- | --- | --- | --- | ---
3 | optimize-php-file-parsing | 4m 46s | N/A | https://blackfire.io/profiles/234d548a-38cf-4be8-8c0e-5bb06c7bf9a8/graph
3 | master | 4m 49s | N/A | https://blackfire.io/profiles/cf557e4c-1dc9-4286-b532-7a7e0c50c6d0/graph
4 | optimize-php-file-parsing | 4m 48s | N/A | https://blackfire.io/profiles/e84c74c9-2d19-43a7-b3c3-9ea59ffc078c/graph
4 | master | 4m 49s | N/A | https://blackfire.io/profiles/9bfe0064-719d-480a-ac69-03887ebb1727/graph
5 | optimize-php-file-parsing | 4m 52s | N/A |  https://blackfire.io/profiles/1bda8e63-f6ef-4b11-bf97-c1b3585fd9fb/graph
5 | master | 5m 01s | N/A | https://blackfire.io/profiles/7492d247-b821-4f0d-8bc6-bfe8243e543e/graph

Comparing profile (from `optimize-php-file-parsing` to `mater`): https://blackfire.io/profiles/compare/c00d45e4-6f6b-4678-9a49-1f85b441c943/graph

#### Without Blackfire, using unix `time`:

№ of run | branch | time 
--- | --- | --- 
1 | optimize-php-file-parsing | 3m 58s 
1 | master | 4m 01s
2 | optimize-php-file-parsing | 3m 55s
2 | master | 4m 01s

--------
So it looks like when xdebug is enabled, serializing slows down execution and adds more memory consumption (which is expected).

When xdebug is automatically disabled (master branch), serializing works faster for several seconds.

I don't know how to interpret results with xdebug enabled and why it totally different from what we have with xdebug disabled.

Since we already have "disabling xdebug" in master, we can merge this branch. Or, alternatively, we can implement the second idea with cloning nodes (Nikic mentioned it as well) and compare it with serializing.

**Comment by theofidry** [comment]
I think the comparison should be the other way around: from master to optimize-php-file-parsing. We can see we have ~1000 PHP-Parser call less so I guess that's a win even though I expected a much bigger diff to be honest.

**Comment by borNfreee** [comment]
> I think the comparison should be the other way around: from master to optimize-php-file-parsing

https://blackfire.io/profiles/compare/593b33df-9566-4fa9-be52-4213f775e401/graph with xdebug enabled

**Comment by theofidry** [comment]
xdebug or not it looks fishy we can -X numbers of calls on the parser but the exact same +X on the diff (which somehow cost more). I can't see what's causing it in the code, will have to have a deeper look

**Comment by borNfreee** [comment]
From [this profile](https://blackfire.io/profiles/d216a715-22c9-409b-adf7-a194acb196c6/graph?settings%5Bdimension%5D=wt&settings%5Bdisplay%5D=landscape&settings%5BtabPane%5D=nodes&selected=PhpParser%5CParserAbstract%3A%3Aparse&callname=main()) (master, without `serialize`, [screenshot](http://take.ms/q91JQ)) we can see that `MutationCreator::create` is called X times, and inside this function master branch also calls `differ->diff()` X times.

Considering the following [paragraph](https://blackfire.io/docs/reference-guide/analyzing-call-graphs#beyond-the-hot-paths) from the Blackfire documentation

> Note that part of the data preprocessing done by blackfire involves removing (aka. pruning) function calls that take less that 1% of the global costs on all dimensions (time, cpu, memory, etc.). This is done to reduce the size of the profiles and to provide a responsive UI. You won't be able to find these removed nodes in the call graph nor in the functions list.

I think `differ->diff` is hidden (not included) in the FROM profile, but exists in the TO profile, that why we have 0 to X difference.

From the code, I can't see `diff()` method is called more than in master, so I believe it's falsy blackfire result

**Comment by theofidry** [comment]
Hm maybe it is but that don't explain how come the diff is suddenly much more expansive. Maybe we could try on a smaller project where we can have multiple samples for a more reliable result, one shot profiles are easily wanky.

**Comment by borNfreee** [comment]
> Or, alternatively, we can implement the second idea with cloning nodes (Nikic mentioned it as well) and compare it with serializing.

Here is an implementation of cloning idea: https://github.com/infection/infection/commit/b3dfda8fcaf25f44d8e6b3947126c7937d443dd5

Benchmark:

#### Without "disabling xdebug" feature:

№ of run | branch | time | memory usage | blackfire profile
--- | --- | --- | --- | ---
1 | clone-ast | 7m 35s | 200MB | https://blackfire.io/profiles/d4971783-6d8d-484f-a199-8189a5cbacb4/graph
1 | master | 9m 4s | 200MB | https://blackfire.io/profiles/4493cf31-d9ba-4816-b73b-6613a9ea9f25/graph

Comparing profile (from `master` to `clone-ast`: https://blackfire.io/profiles/compare/d75c0a0a-9c4b-4dcf-8061-d1d2269d75fe/graph

#### With "disabling xdebug" feature:

№ of run | branch | time | memory usage | blackfire profile
--- | --- | --- | --- | ---
3 | clone-ast | 3m 19s | N/A | https://blackfire.io/profiles/6523045f-ed4a-43aa-9fa2-ff757278781c/graph
3 | master | 3m 28s | N/A | https://blackfire.io/profiles/a2dfaf16-12d8-4456-b676-8941691e283a/graph


--------

So, cloning the AST seems gives a good perf boost both with xdebug on and off.

**Comment by sidz** [comment]
@borNfreee vote for your solution. But can we parse file just one time? right now we're doing it two places:
in the MutationsGenerator and MutantCreator. Can we try to save original AST in the Mutation?

**Comment by borNfreee** [comment]
@sidz implemented in #95

**Comment by borNfreee** [comment]
Merged https://github.com/infection/infection/commit/44cd016a9f72d56ff3168eb2552ae435ee9f1d46
